pkg/sdkclient/batchmapper/client.go

```diff
@@ -0,0 +1,175 @@
+	// errors in sending, UDF errors etc
+	// These are propagated to the applier for further handling
+	errCh := make(chan error)
+
+	// response channel for streaming back the results received from the gRPC server
+	// TODO(map-batch): Should we keep try to keep this buffered?
+	responseCh := make(chan *batchmappb.BatchMapResponse)
+
+	// BatchMapFn is a bidirectional streaming RPC
+	// We get a Map_BatchMapFnClient object over which we can send the requests,
+	// receive the responses asynchronously.
+	// TODO(map-batch): this creates a new gRPC stream for every batch,
+	// it might be useful to see the performance difference between this approach
+	// and a long-running RPC
+	stream, err := c.grpcClt.BatchMapFn(ctx)
+	if err != nil {
+		go func() {
+			errCh <- sdkerr.ToUDFErr("c.grpcClt.BatchMapFn stream", err)
+		}()
+		return responseCh, errCh
+	}
+
+	// read the response from the server stream and send it to responseCh channel
+	// any error is sent to errCh channel
+	go func() {
+		// close this channel to indicate that no more elements left to receive from grpc
+		// We do defer here on the whole go-routine as even during a error scenario, we
+		// want to close the channel and stop forwarding any more responses from the UDF
+		// as we would be replaying the current ones.
+		defer close(responseCh)
+
+		var resp *batchmappb.BatchMapResponse
+		var recvErr error
+		index := 0
+		for {
+			select {
+			// handle a context done
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			default:
+				resp, recvErr = stream.Recv()
+				// check if this is EOF error, which indicates that no more responses left to process on the
+				// stream from the UDF, in such a case we return without any error to indicate this
+				if errors.Is(recvErr, io.EOF) {
+					return
+				}
+				// If this is some other error, propagate it to error channel,
+				// also close the response channel(done using the defer close) to indicate no more messages being read
+				errSDK := sdkerr.ToUDFErr("c.grpcClt.BatchMapFn", recvErr)
+				if errSDK != nil {
+					errCh <- errSDK
+					return
+				}
+				// send the response upstream
+				responseCh <- resp
+				index += 1
+			}
+		}
+	}()
+
+	// Read from the read messages and send them individually to the bi-di stream for processing
+	// in case there is an error in sending, send it to the error channel for handling
+	go func() {
+		// TODO(map-batch): Should we check for ctx.Done here as well?
+		for inputMsg := range inputCh {
+			err = stream.Send(inputMsg)
+			if err != nil {
+				errCh <- sdkerr.ToUDFErr("c.grpcClt.BatchMapFn", err)
+				// TODO(map-batch): Check if we should still do a CloseSend on the stream even if we have
+				// received an error. Ideally on an error we would be stopping any
+				// further processing and go for a replay so this should not be required.
+				// but would be good to verify.
+				// Otherwise this would be a break
+				return
```

Comment:
We don't need to again call `closeSend()`, if there are any errors `gRPC` stream will be closed. Please remove the TODO